### DISCUSSION TITLE
1.4.0_RC1

### DISCUSSION THREAD

**Discussion by francois-normandin** [JKISoftware/Caraya#172] [General]
This release introduces a new Test or Assert **Property** feature that developers can use to add attributes to their individual tests or assertions. The attributes are in the form of key-value pairs and are added to an Extended Test Result cluster to be processed by custom report classes. The feature is optional and accessible through a new Assert/Test method output terminal, as to not have any impact on test execution speeds.

The Test Manager was also modified to use a class-based messaging strategy, as opposed to a variant that can be interpreted based on the name. This feature is to prepare the Test Manager decoupling in future versions, but was also a good opportunity to implement the above-mentioned properties in a more scalable way. The Test and Assert Nodes use extensions to the **TestManagerMsg** base class to broadcast to the Test Manager. Properties are extensions of this base class as well.

Finally, the Test Manager UI now has a Text-based report view in addition to the customary tree structure. It is an experiment to gather feedback on how Caraya users engage with the UI.

### Release Notes:
[Fix: 169] Test Runner now filters the Excluded VI List when discovering tests
[Fix: 166] Test Runner linkage info now errors on invalid/broken test during discovery
[Fix: 164] Inverted inputs in Assert Almost Equal are now consistent with other assertions
[Fix: 163] Linkage Info now skips conditionally disabled methods
[Feature: 161] XML JUnit can now be localized (defaults to system's regional settings)
[Fix: 160] Space character in test label now has no impact on result tree location
[Feature: 158] Verbose explanations can now be forced even for tests that assert
[Fix: 155] Conditionally disabled methods do not produce bad linkage in Test Runners
[Feature: 152] Test manager generates event as each individual tests and assertions are processed
[Feature: 151] Verbose explanations can now be added to any assertion
[Feature: 148] Failure descriptions can now be added to any report, including JUnit

###Contributors
Thanks for all the contributions that led to this release, whether by reporting bugs, requesting features that made it to release or by lending a hand in fixing issues or releasing features:

Tanner Blair (@tannerblair)
Sébastien Michaud (@SebMichaud)
Ajay MV (@digiajay)
Francois Normandin (@francois-normandin)
Roberto Olea (@Robbie20)
Hunter Smith (@HunterJKI)
(@phollyni)

<hr /><em>This discussion was created from the release <a href='https://github.com/JKISoftware/Caraya/releases/tag/1.4.0_RC1'>1.4.0_RC1</a>.</em>

**Comment by francois-normandin** [comment]
### New Feature: Report Tab

I would like to have feedback on the addition of the second tab (Report).

It displays only a **Simple Text Report** of the tests in this release. The target would be to make this tab be configurable by the user, so that another type of report (JUnit or custom) could be used instead. Is the workflow agreeable with the community or does this provide very little value as it is? What are the feature that would make this tab of interest? Copy to Clipboard support might be one of them. Anything else? 

![image](https://github.com/JKISoftware/Caraya/assets/11728548/3d5cdf41-1022-46c2-a84a-b75dde01ab67)

**Comment by francois-normandin** [comment]
### New feature: Property palette

The main feature of this release is the addition of a Property palette in Caraya`s Advanced palette.
![image](https://github.com/JKISoftware/Caraya/assets/11728548/05e5d474-11f1-4073-9cbd-954166deb592)

Properties are essentially Key-Value pairs that can be automatically tied with a registered test or assertion.
![image](https://github.com/JKISoftware/Caraya/assets/11728548/841d6522-1436-44b9-a8dc-b3bae4f1a66c)

Each Test or Assert node now outputs are Property class (Property.Assert, Property.Test) which encapsulates the unique identifier. Adding properties is as simple as wiring the output class to a property method.

![image](https://github.com/JKISoftware/Caraya/assets/11728548/f252e3ee-e2d0-4289-9c70-f427cfd3d5a6)

There are Test-specific properties: **Hyperlink, RequirementID**.
There are Assert-specific properties: **Measurement and Verbose Explanations**
There are Generic properties: **Key-Value Pairs**

Those classes can be extended by the developer to provide standardized reuse property nodes for a project or a company`s reuse code.

What are the types of properties that the community would like to benefit from? The Key-Value pair can be used in any context, but perhaps there might be more universal properties that should be supported in the palette? Tell us about your needs or how you would see yourself using those extra functionalities.

Check out the Examples palette: 
![image](https://github.com/JKISoftware/Caraya/assets/11728548/30f71254-c5c1-44ed-99cd-5efc67f8ecf0)

**Reply by SebMichaud** [reply]
Hi there,
Fantastic work, thanks.
When will it be available on vipm ?
Thanks

**Reply by francois-normandin** [reply]
I don't have a fixed date for making it fully accessible in VIPM. This pre-release period is intended to make sure it can be installed and tested by a few people that will confirm it does not break their already established workflows and CI pipelines.

I don't think it introduced any breaking changes.
I can tell you that, no matter what, these features are here to stay. They will not be stripped out like it's always possible during a beta testing period. It's only a minor release addressing a few requests and bug fixes.

I'd say it will be made available on VIPM in early July if there are no show stoppers being reported. I'm about to test it in production myself, this week.

**Comment by francois-normandin** [comment]
### New Feature: Event Report

Extension of the report class to natively support non-file-based reporting. Individual Assertion events are published in JSON format as they are processed by the Test Manager. A user-defined user can be injected in a report instance and injected into a Test Suite like any file-based report class. 

![image](https://github.com/JKISoftware/Caraya/assets/11728548/71430267-5280-412e-aad8-643c634a6d81)
